src/node/transaction.cpp

```diff
@@ -50,12 +50,14 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
     if (!node.mempool->exists(hashTx)) {
         // Transaction is not already in the mempool.
         TxValidationState state;
-        CAmount fee{0};
         if (max_tx_fee) {
-            // First, call ATMP with test_accept and check the fee. If ATMP
-            // fails here, return error immediately.
-            if (!AcceptToMemoryPool(*node.mempool, state, tx,
-                nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee)) {
+            // First, check that transaction fee does not exceed max_tx_fee.
+            Optional<CAmount> fee = GetTransactionFee(*node.mempool, state, *tx);
+            if (fee == nullopt) {
+                // Failed GetTransactionFee indicates a validation failure.
+                // Use ATMP with test_accept to return the validation error.
+                AcceptToMemoryPool(*node.mempool, state, tx,
+                    nullptr /* plTxnReplaced */, false /* bypass_limits */, true /* test_accept */);
```

Comment:
I may be missing something, but I think this can be simplified further by dropping the call to ATMP here. Suppose the transaction has _two_ problems, one that `GetTransaction()` catches and another that's outside its scope. `GetTransactionFee()` will return a failure and presumably set `state` appropriately (I didn't verify this, but if it doesn't, that's a bug that should be fixed). I don't see what's wrong with just returning that error. 

It's true that by calling ATMP here, `state` may get set to a _different_ error, but it's not clear to me why that's a better error to return.

Reply:
ATMP should be the authority on validation; I don't want to trust anything else with filling in `TxValidationResult`. I think I could make it a little cleaner and have `GetTransactionFee` just not touch state, but `CheckTxInputs` needs it 🤔

Reply:
Tcb